### ISSUE TITLE
Add some rough resource requirements for memory intensive files

### ISSUE THREAD

**Issue by bgamari** [snowleopard/hadrian#294]
One of the great benefits of Hadrian will be the ability to control parallelism to ensure smooth parallel builds on smaller machines. For instance, my ARM builder is a four-core machine (eight if one includes the smaller A7 cores) but only 2GB of memory. Unfortunately there are a few files in the build which require an extraordinary amount of memory to compile (e.g. `DynFlags`) so using even `-j3` on this machine will invariably result in the swap-death of the machine.

It would be great if we could start adding some rough resource annotations to the build system to reflect these choke points. Or, even better, if we could instrument Hadrian to use previously recorded resource requirements to inform the build.

**Comment by snowleopard** [comment]
We already use Shake resources in Hadrian for limiting concurrency when mutating the package database, so the infrastructure for passing resources around is already there.

See https://github.com/snowleopard/hadrian/blob/master/src/Rules.hs#L56-L62, for example.

Adding a memory resource is fairly straightforward. Would you like to try this? I'm happy to help. I don't expect any difficulties, but you have a better understanding of what exactly you'd like to achieve, so I think you could do this faster.

> Or, even better, if we could instrument Hadrian to use previously recorded resource requirements to inform the build.

I believe Shake already does something like this under the hood, but I don't think Shake has any API to expose this sort of profiling information to users...

**Comment by bgamari** [comment]
> Adding a memory resource is fairly straightforward. Would you like to try this? I'm happy to help. I don't expect any difficulties, but you have a better understanding of what exactly you'd like to achieve, so I think you could do this faster.

One option would be to maintain a global `Map FilePath Int` which we can centrally define somewhere which gives a rough peak RSS of GHC while compiling each source file in the tree. This map could be read from a file, which at some point in the future we could teach Hadrian to generate.

The problem that I see is that the memory resource should be a global resource and consequently would need to be threaded throughout the build system. I'm not sure we'd want to do this manually as is done in the example that you cite.

> I believe Shake already does something like this under the hood, but I don't think Shake has any API to expose this sort of profiling information to users...

I suspect we will need a bit more than what Shake provides since I doubt it tracks, for instance, maximum memory residency.

**Comment by snowleopard** [comment]
> The problem that I see is that the memory resource should be a global resource and consequently would need to be threaded throughout the build system. I'm not sure we'd want to do this manually as is done in the example that you cite.

A map like `Map FilePath Int` could be made available in `Action` monad using `shakeExtra`, which is part of `ShakeOptions` and can therefore be obtained by `getShakeOptions`. As a first simple solution, we could return 1 for most source files, apart from a few problematic ones wew know about.

> I suspect we will need a bit more than what Shake provides since I doubt it tracks, for instance, maximum memory residency.

Indeed. I believe Shake only tracks the time it takes to run each build rule.

**Comment by ndmitchell** [comment]
Shake tracks time spent in each rule, but nothing more. For the resource of
memory you probably want the peak memory used by a child group of
subprocesses, which cmd/command don't give you - and I'm not sure is in the
Haskell process library either. Using information about past memory usage
to do scheduling decisions seems valuable, but not really possible in
Shake. Using that information to apply resources to future compilations
doesn't seem such a great idea - I worry you're going to be crashing in the
first iteration and after any clean build. If most compilations require
1Gb+ then limit every GHC process. If it's only a handful, better to have a
whitelist, perhaps with a check after the fact.

Unfortunately throttling memory usage is just a series of heuristics,
nothing really concrete :(

**Comment by snowleopard** [comment]
We could turn resources into expressions:
```haskell
type Resources = DiffExpr [(Resource, Int)]

-- We currently have these resource settings mixed with top-level Rules.hs definitions
defaultResources :: Resources
defaultResources = mconcat [ builder GhcPkg ? append writePackageDb
                           , builder Ghc    ? append readPackageDb
                           , ... ]

resources :: Resources 
resources = defaultResources
```

Implementing this should be straightforward. The user will then be able to tweak Hadrian very flexibly:
```haskell
resources :: Resources 
resources = defaultResources <> userResources

userResources :: Resources
userResources = [ input "//DynFlags.*" ? append [(memory, 1000)]
                , stage1 ? builder (Ghc Link) ? append [(disk, 1), (memory, 100)]
                , ... ]
```

@bgamari @ndmitchell What do you think? This is not an automated solution, but will probably allow Hadrian users to relatively easily tailor GHC build to their system constraints.

**Comment by ndmitchell** [comment]
Sounds quite reasonable - although might be overkill - do people really need to tweak resources that often?

**Comment by snowleopard** [comment]
I agree this seems excessive at first sight, but if the user does want to tweak resources even for just a single file like `DynFlags` then it would be very nice if they could do it in `UserSettings.hs` without the need to understand how the rest of Hadrian works, which leads to adding another type of expressions.

**Comment by bgamari** [comment]
Andrey Mokhov <notifications@github.com> writes:

>
> @bgamari @ndmitchell What do you think? This is not an automated
> solution, but will probably allow Hadrian users to relatively easily
> tailor GHC build to their system constraints.
>
Sounds like a reasonable start to me.
